src/builder/QuarkBuilder.sol

```diff
@@ -1100,25 +1139,35 @@ contract QuarkBuilder {
      * the destination chain (once bridging is complete) will have a sufficient
      * amount of the payment token to cover the non-bridged actions.
      *
-     * Optional `supplementalPaymentTokenBalance` param describes an amount of
-     * the payment token that might have been received in the course of an
-     * action (for example, withdrawing an asset from Compound), which would
-     * therefore not be present in `chainAccountsList` but could be used to
-     * cover action costs.
+     * Optional `paymentTokenReceived` and `paymentTokenSpent` params describe
+     * an amount of the payment token that might have been spent or received in
+     * the course of an action (for example, withdrawing an asset from Compound
+     * or repaying a borrow in Compound), which would therefore not be present
+     * in `chainAccountsList`.
      */
     function assertSufficientPaymentTokenBalances(
         Actions.Action[] memory actions,
         Accounts.ChainAccounts[] memory chainAccountsList,
         uint256 targetChainId
     ) internal pure {
-        return assertSufficientPaymentTokenBalances(actions, chainAccountsList, targetChainId, 0);
+        return assertSufficientPaymentTokenBalances(actions, chainAccountsList, targetChainId, 0, 0);
+    }
+
+    function assertSufficientPaymentTokenBalances(
```

Comment:
The changes to `assertSufficientPaymentTokenBalances` are the biggest change in this PR.

The structure of this function assumes that you can determine how much of the payment token is spent in an action based on its `ActionContext`. Frustratingly, that's not the case with a max Repay, because the ActionContext's `amount` is `type(uint256).max`.

So I've refactored this function to not read values from the various ActionContexts but instead to take in `paymentTokenReceived` and `paymentTokenSpent` arguments, so the caller can pass in their calculation of how much of the payment token was gained or lost in the actions.

The one part of the calculation that I've not recreated in this refactoring is the amount of the payment token that is spent and/or received by any Wrapping/Unwrapping actions that are added. Technically, I think we could do it, but it seems unnecessary as I don't think we're going to be doing any wrapping/unwrapping of USDC.

Reply:
I'm open to other suggestions if this refactor seems too dramatic.

Reply:
hmm yeah, and it feels like part of the `assertSufficientPaymentTokenBalance` logics gets removed outside of the function and scattered into various places in above code blocks.  Would prefer to make that function to covers most cases, so people can simply put that check function at the end assuming it can help reviewing the final answer and tell you if you can afford it or not. 

If all action needs to calculate paymentToken on their own, then probably `wrapOrUnwrap` needs to be refactor in case of wrap/unwrap away the paymentToken cases.

Wondering if it's hard to replicate the math done in line166 in this function to calculate here instead 🤔

Reply:
We can't do the math here with the existing arguments, because there is no piece of data in `chainListAccounts` or `actions` that captures what the actual repayAmount is going to be. The `amount` argument in the `ActionContext` is just `uint256.max`.

The user's comet positions are available as part of `chainListAccounts`, but you don't know which account to look for the max borrow amount for.

Reply:
I see.